capstone/fitness/src/main/java/com/google/sps/servlets/ProgressModelServlet.java

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import com.google.sps.fit.*;
+import com.google.sps.progress.*;
+import com.google.sps.util.*;
+import com.google.sps.util.Metadata.Sort;
+import java.io.IOException;
+import java.util.*;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/pro")
+public class ProgressModelServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String goalStepsJson = DataHandler.getGoalSteps();
+    List<JsonExercise> display = null;
+    if(goalStepsJson != null) {
+      // Replicate existing model and stored version based on session in DataHandler.
+      ProgressModel model = new ProgressModel.Builder()
+                  .setJsonGoalSteps(goalStepsJson)
+                  .build();
+      if(model.updateModel()) {
+        DataHandler.setGoalSteps(model.toJson());
+      }
+      display = getProgressDisplays(paginate(model.toArray(), request.getSession()));
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(getJson(display));
+  }
+
+  private GoalStep[] paginate(GoalStep[] goalSteps, HttpSession session) {
```

Comment:
I thought pagination also included filters, however I am filtering it then getting the desired page out of it. I'm not really sure if I can put them into separate methods because a lot of the variables associated with the filters and pagination overlap to create the entire display. The filtering section is under the comment "Get split array based on filter" and the page finder section is under the comments "Modify where the split array starts based on page number and goal step count per page." and "Add visible goal steps based on metadata parameters."

Also the initialization and decrement of the startingIndex variable handles the case where the filter is changed in the middle of a page. If it doesn't have a page for that filter, then it decrements the page until it finds an appropriate page.